### PR TITLE
Use EventsCollector when testing commands #504

### LINKED ISSUE #504: Use EventsCollector when testing commands
`HelpCommandTest`, `ExitCommandTest` and `SelectCommandTest` manually subscribes to the event that we are expecting and handle the event ourselves.

`EventsCollector` is a test helper utility class that does the same thing without having us to subscribe manually.

Let's use `EventsCollectors` for our test.

### PR DESCRIPTION
**Pull request by yamgent**
Fixes #504.

Proposed merge commit message:
```
Some command tests check that the correct events are being triggered by
manually subscribing to the events that they are expecting.

We already have EventsCollector class in our codebase that catches all
possible events, and allows us to extract data from the triggered
events as well.

Let's replace all manual subscription of events with EventsCollector
instead.
```

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@yamgent submitted v1 for review.

* [v1 1/3] [HelpCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/12f8316f21c2f47b0d85df0846f31f33669bb880)
* [v1 2/3] [ExitCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/311f5261697597c1a6b43394279ef0a4d7bbf4d5)
* [v1 3/3] [SelectCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/f534768607bb382e21c66724ef5c20894fb2c5f9)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...f534768607bb382e21c66724ef5c20894fb2c5f9))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/507/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/logic/commands/SelectCommandTest.java [hunk 1]
```diff
@@ -98,31 +87,30 @@ private void showFirstPersonOnly(Model model) {
...
-        eventTargetedJumpIndex = null;
-
         SelectCommand selectCommand = prepareCommand(index);
 
         try {
             selectCommand.execute();
             fail("The expected CommandException was not thrown.");
         } catch (CommandException ce) {
             assertEquals(expectedMessage, ce.getMessage());
-            assertNull(eventTargetedJumpIndex);
```
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/logic/commands/SelectCommandTest.java]: we shouldn't remove this check. Perhaps we should improve the API of `EventsCollector` by implementing an `isEmpty()` method.

**Comment by Zhiyuan-Amos** [comment]
Another comment: Can we improve the API of `EventsCollector` by implementing a `getFirst()` (sidenote: we should also remove the `reset()` method) or something like that. Especially since this is a Test Utility class, and the methods implemented don't have to be general; it simply needs to meet the needs of the making the test cases easier to implement. So the generic `get(int index)` method should be replaced by `get()` and perhaps we can change the internal implementation to use a single `BaseEvent` instead of `List<BaseEvent>`, since in our test cases, we are only using `get(0)`.

File: src/test/java/seedu/address/logic/commands/ExitCommandTest.java [hunk 2]
```diff
@@ -4,31 +4,19 @@
...
-    }
 
     @Test
     public void execute_exit_success() {
+        EventsCollector eventCollector = new EventsCollector();
+
         CommandResult result = new ExitCommand().execute();
         assertEquals(MESSAGE_EXIT_ACKNOWLEDGEMENT, result.feedbackToUser);
-        assertTrue(isEventCaught);
+        assertTrue(eventCollector.get(0) instanceof ExitAppRequestEvent);
```
Comment (limmlingg) [inline src/test/java/seedu/address/logic/commands/ExitCommandTest.java]: `.get(0)` doesn't sound intuitive. Can we implement another method in EventCollector or in the tests to make this more understandable?

**Comment by limmlingg** [comment]
Merge commit message:

Some command tests **check** that the correct events are being triggered ~They do so~ by manually subscribing to the events that they are expecting.

We already have EventsCollector class in our codebase that catches all possible events, and **allows** us to extract data ~out of~ **from** the triggered **events** as well.

Let's replace all manual subscription of events ~by using~ **with** EventsCollector instead.

**Comment by yamgent** [comment]
> (sidenote: we should also remove the reset() method)

Should be a separate PR (can even be a first timer issue).

**Comment by CanIHasReview** [comment]
# v2

@yamgent submitted v2 for review.

* [v2 1/5] [EventsCollector: Add getMostRecent\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/bb2e414de43329d7a947a19e4eef91847bbf44ed)
* [v2 2/5] [HelpCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/4d4ddb65b45a04b466f0607a2ae98303b0a877b4)
* [v2 3/5] [ExitCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/eabbea7995dfffec2aa1a4c9c184940dbe36d844)
* [v2 4/5] [EventsCollector: Add isEmpty\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/4d58dac8624253d3eb7691990de8d907005b0422)
* [v2 5/5] [SelectCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/f3c3e9f4c0c27a4c4fde5fa1cf40807f79ca1178)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/48775e45f9e37d5c00b99f7241d2856b3e2eac9e...f3c3e9f4c0c27a4c4fde5fa1cf40807f79ca1178)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/e1866dfa0d2812202d6fc8fbd69253c9f5870c89~1...e1866dfa0d2812202d6fc8fbd69253c9f5870c89))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/507/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/test/java/seedu/address/testutil/EventsCollector.java [hunk 3]
```diff
@@ -39,4 +39,15 @@ public void reset() {
     public BaseEvent get(int index) {
```
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/testutil/EventsCollector.java]: i think this get method is unused after you have updated the tests with the newer `getMostRecent()`.

**Comment by CanIHasReview** [comment]
# v3

@yamgent submitted v3 for review.

* [v3 1/6] [EventsCollector: Add getSize\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/b8f0539ac2fd52141ef20fe7625886929a560bcd)
* [v3 2/6] [EventsCollector: Add getMostRecent\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/fada53bba8d233019ca34303511090e9b18ba99d)
* [v3 3/6] [HelpCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/55b7180eda631ad8c5ad8a57d391ce4c81783efb)
* [v3 4/6] [ExitCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/9258e653944cec8a2a3c3542b11703c22ce50c21)
* [v3 5/6] [SelectCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/70165f449f90365be171927416894b67e1e3c866)
* [v3 6/6] [EventsCollector: Remove get\(int\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/ec0f7dd4c4cccdab26116cd513c6992ad8d2c5d7)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/48775e45f9e37d5c00b99f7241d2856b3e2eac9e...ec0f7dd4c4cccdab26116cd513c6992ad8d2c5d7)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/acf3d8cf0457f5db9873ac228182e11202649fda~1...acf3d8cf0457f5db9873ac228182e11202649fda))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/507/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by yamgent** [comment]
isEmpty(), reset on assertCommandSuccess/assertCommandFailure (or just create event collector in method?)

**Comment by CanIHasReview** [comment]
# v4

@yamgent submitted v4 for review.

* [v4 1/7] [EventsCollector: Add getSize\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/b8f0539ac2fd52141ef20fe7625886929a560bcd)
* [v4 2/7] [EventsCollector: Add getMostRecent\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/fada53bba8d233019ca34303511090e9b18ba99d)
* [v4 3/7] [HelpCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/55b7180eda631ad8c5ad8a57d391ce4c81783efb)
* [v4 4/7] [ExitCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/9258e653944cec8a2a3c3542b11703c22ce50c21)
* [v4 5/7] [EventsCollector: Add isEmpty\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/b5bdd794bfe8e1b0adaa2e54baf8d809f46b2e3e)
* [v4 6/7] [SelectCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/e6e9d4e8e01e34a0faa7a811c821260feb2bd94e)
* [v4 7/7] [EventsCollector: Remove get\(int\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/3d1763cfddf35fbbf461940f17d756b7b0a3d4d8)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/48775e45f9e37d5c00b99f7241d2856b3e2eac9e...3d1763cfddf35fbbf461940f17d756b7b0a3d4d8)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/f6cc55c9d66f873be347fc847779113814066f85~1...f6cc55c9d66f873be347fc847779113814066f85))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/507/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by yamgent** [comment]
> I'm thinking this commit should go after [6/7] since it seems that only SelectCommandTest uses it? If we move it, the commit message should also change to mention SelectCommandTest.

Make sense, will move it down.

**Comment by CanIHasReview** [comment]
# v5

@yamgent submitted v5 for review.

* [v5 1/7] [EventsCollector: Add getSize\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/943d1286f93d140b62e0ecf3f2c837a2ebb7ac4b)
* [v5 2/7] [EventsCollector: Add getMostRecent\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/0fd010a8095baa15b79f6169466f92d151151de5)
* [v5 3/7] [HelpCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/36edb5221a2ea76f8136941db55cf869635c426f)
* [v5 4/7] [ExitCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/21b0e337c24db98d1fe8d2580fb411f18e812006)
* [v5 5/7] [SelectCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/64fd5b9fa247f3bff113b474409d0ea934655336)
* [v5 6/7] [EventsCollector: Add isEmpty\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/e8f82122b613a592a31ecd8eee119342f636bc2d)
* [v5 7/7] [EventsCollector: Remove get\(int\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/123ff251b2a4c0d96cb9d3fe85e3e428595589fb)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/8cfab29d930e3fa60c7411a7eb6992373cff0686...123ff251b2a4c0d96cb9d3fe85e3e428595589fb)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/a50eeaf8331cb06c85c82ff6b6457885823f6bba~1...a50eeaf8331cb06c85c82ff6b6457885823f6bba))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/507/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v6

@yamgent submitted v6 for review.

* [v6 1/7] [EventsCollector: Add getSize\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/943d1286f93d140b62e0ecf3f2c837a2ebb7ac4b)
* [v6 2/7] [EventsCollector: Add getMostRecent\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/0fd010a8095baa15b79f6169466f92d151151de5)
* [v6 3/7] [HelpCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/36edb5221a2ea76f8136941db55cf869635c426f)
* [v6 4/7] [ExitCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/21b0e337c24db98d1fe8d2580fb411f18e812006)
* [v6 5/7] [SelectCommandTest: Use EventsCollector to receive events](https://github.com/se-edu/addressbook-level4/pull/507/commits/64fd5b9fa247f3bff113b474409d0ea934655336)
* [v6 6/7] [EventsCollector: Add isEmpty\(\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/1bf9b4f4176f71db4d19f30641182c6670b0e42d)
* [v6 7/7] [EventsCollector: Remove get\(int\)](https://github.com/se-edu/addressbook-level4/pull/507/commits/fbdc48389e26e64d9089a274e243ef565b3ee7d1)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/8cfab29d930e3fa60c7411a7eb6992373cff0686...fbdc48389e26e64d9089a274e243ef565b3ee7d1)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/0948af8b4ba7f8f06422efc8105f6988ee2e6327~1...0948af8b4ba7f8f06422efc8105f6988ee2e6327))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/507/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Pull request merged**